pytest/tests/loadtest/locust/common/social.py

```diff
@@ -0,0 +1,279 @@
+            "set_status", args, 100 * TGAS, 0)
+
+        # Batch the two actions above into one transaction
+        nonce = self.contract.use_nonce()
+        key = self.contract.key
+        return transaction.sign_and_serialize_transaction(
+            key.account_id, nonce, [call_new_action, call_set_status_action],
+            block_hash, key.account_id, key.decoded_pk(), key.decoded_sk())
+
+
+class InitSocialDbAccount(Transaction):
+    """
+    Send initial storage balance to ensure the account can use social DB.
+
+    Technically, we could also rely on lazy initialization and just send enough
+    balance with each request. But a typical user sends balance ahead of time.
+    """
+
+    def __init__(self, contract_id: str, account: Account):
+        super().__init__()
+        self.contract_id = contract_id
+        self.account = account
+
+    def sign_and_serialize(self, block_hash):
+        args = json.dumps({"account_id": self.account.key.account_id})
+        return transaction.sign_function_call_tx(self.account.key,
+                                                 self.contract_id,
+                                                 "storage_deposit",
+                                                 args.encode('utf-8'),
+                                                 300 * TGAS, 1 * NEAR_BASE,
+                                                 self.account.use_nonce(),
+                                                 block_hash)
+
+
+def social_db_build_index_obj(key_list_pairs: dict) -> dict:
+    """
+    JSON serializes the key - value-list pairs to be included in a SocialDB set message.
+
+    To elaborate a bit more, SocialDB expects for example
+    ```json
+    "index": { "graph": value_string } 
+    ```
+    where value_string = 
+    ```json
+    "[{\"key\":\"follow\",\"value\":{\"type\":\"follow\",\"accountId\":\"pagodaplatform.near\"}}]"
+    ```
+    So it's really JSON nested inside a JSON string.
+    And worse, the nested JSON is always a list of objects with "key" and "value" fields.
+    This method unfolds this format from a leaner definition, using a dict to
+    define each `value_string`.
+    """
+    obj = {}
+    for index_key, values in key_list_pairs.items():
+        unfolded_list = [{"key": k, "value": v} for k, v in values.items()]
+        obj[index_key] = json.dumps(unfolded_list)
+    return obj
+
+
+def social_db_set_msg(sender: str, values: dict, index: dict) -> str:
+    """
+    Construct a SocialDB `set` function argument.
+
+    The output will be of the form:
+    ```json
+    {
+      "data": {
+        "key1": value1,
+        "key3": value2,
+        "key4": value3,
+        "index": {
+          "index_key1": "[{\"index_key_1_key_A\":\"index_key_1_value_A\"}]",
+          "index_key2": "[{\"index_key_2_key_A\":\"index_key_2_value_A\"},{\"index_key_2_key_B\":\"index_key_2_value_B\"}]",
+        }
+      }
+    }
+    ```
+    """
+    updates = values
```

Comment:
Note, that assignment doesn't create a copy of an object, and the mutation of `updates` will lead to the mutation of the input parameter `values` and will be visible outside of this function. If you want to do a real copy, you need to do `updates = values.copy()`.

Example:
```python
>>> a = {1: "hello"}
>>> b = a
>>> b[2] = "world"
>>> b
{1: 'hello', 2: 'world'}
>>> a
{1: 'hello', 2: 'world'}
```

Reply:
Ah, I didn't think much about the mutation of `values`, it doesn't matter much, at least the way it's currently used... But I like the `.copy()` addition a lot here, mutating the incoming value seems just wrong.